---
title: "LowKi: kinship matrix from low depth data"
author: "Anthony Herzig and Hervé Perdry"
date: "`r Sys.Date()`"
output: html_document
vignette: >
  %\VignetteIndexEntry{LowKi}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r message=FALSE, warning=FALSE, echo = FALSE}
library("knitr")
```

LowKi is a package for computing Kinship and Fraternity matrices from shallow sequencing data. It uses vcf files with 'PL' or 'GP' fields. 
To install the package, first install `gaston`, `milorGWAS`, `Rcpp`, and `RcppEigen` from the CRAN,
then use `devtools` :

`install.packages(c("Rcpp", "RcppEigen", "gaston", "milorGWAS"))`
`devtools::install_github("genostats/LowKi")`

A prepublication describing the method 
is available on Research Square: [doi:10.21203/rs.3.rs-1109592/v1](https://doi.org/10.21203/rs.3.rs-1109592/v1)

# An example of usage

A small simulated dataset, corresponding to a fictitious small isolated population, is installed with `LowKi`. To build it, five generations of a panmictic population of one thousand individuals per generation were generated by gene dropping; the founders haplotypes were build as mosaics of 1000 genome haplotypes. 
The dataset consists in a 68 Mb vcf file containing shallow data (mean depth 1 to 3) for 200 individuals tkaen from the last two simulated generations, with data in 200,000 autosomal SNPs only – to keep the file reasonably small. Two small rds files are also included, corresponding to the 'true' kinship and fraternity matrices, as computed from the length of genome shared IBD by the simulated individuals.

The path to these three files on your filesystem can be obtained as follows:

```{r message = FALSE}
library(LowKi)
vcf.file <- system.file("extdata", "shallow.vcf.gz", package="LowKi")
kinship.file <-system.file("extdata", "kinship.rds", package="LowKi")
fratern.file <-system.file("extdata", "fraternity.rds", package="LowKi")
```

### Computing the kinship matrix

The function `lowKi` will use the 'PL' field of the vcf file to compute the kinship matrix.

```{r}
K.low <- lowKi(vcf.file)
```

One can also use the hard called genotypes present in the vcf to estimate the kinship matrix (the functions bellow are from package gaston).

```{r results = FALSE}
z <- read.vcf(vcf.file)
K.gen <- GRM(z)
```

Finally, the 'true' kinship values, computed when the dataset was generated, can be loaded by:

```{r}
K.real <- readRDS(kinship.file)
```

Let us compare these values.
```{r fig.width=10, fig.height=5}
par(mfrow=c(1,2))
plot(K.real, K.low, xlab = "true values", ylab = "LowKi estimates")
abline(0,1,col="red")
plot(K.real, K.gen, xlab = "true values", ylab = "Estimates from hard called genotypes")
abline(0,1,col="red")
```

### Computing the fraternity matrix

To compute the fraternity matrix (or 'dominance matrix'), just add `fraternity = TRUE` in `lowKi`.
We compare it again to the 'true' values and to the hard called genotypes estimates.

```{r fig.width=10, fig.height=5}
D.low <- lowKi(vcf.file, fraternity = TRUE)
D.gen <- DM(z)
D.real <- readRDS(fratern.file)

par(mfrow=c(1,2))
plot(D.real, D.low, xlab = "true values", ylab = "LowKi estimates",
     xlim = c(0,0.5), ylim = c(-0.1,0.5))
abline(0,1,col="red")
plot(D.real, D.gen, xlab = "true values", ylab = "Estimates from hard called genotypes", 
     xlim = c(0,0.5), ylim = c(-0.1,0.5)); 
abline(0,1,col="red")
```

# Some available options 

We illustrated the default behavior of the `lowKi` function. We will see below how
to use it further.

### Using 'external' allelic frequencies

The default moment estimates are built on moment estimates from the sample.
One can also provide 'external' allele frequencies through argument `freqs`. 

The function `vcf.allele.freq`
allows to compute Maximum Likelihood estimators of allele frequencies. 
The example below demonstrates its use.

```{r fig.width=10, fig.height=5}
MLfreq <- vcf.allele.freq(vcf.file)
K.low <- lowKi(vcf.file, freqs = MLfreq)
D.low <- lowKi(vcf.file, freqs = MLfreq, fraternity = TRUE)
par(mfrow=c(1,2))
plot(K.real, K.low, xlab = "true values", ylab = "LowKi estimates", main = "kinship matrix")
abline(0,1,col="red")
plot(D.real, D.low, xlab = "true values", ylab = "LowKi estimates", main = "fraternity matrix",
     xlim = c(0,0.5), ylim = c(-0.1,0.5))
abline(0,1,col="red")
```

### The number of individuals used for the adjustment procedure

`LowKi` first computes 'raw' biased estimates, then an adjustment procedure is used to reduce the bias.
The default is to use a small number of individuals for the adjustment. The `adjust.par`
parameter gives three numbers, corresponding to the number of pairs of individuals with the lowest kinship to be used,
the number of pairs with the highest coefficient, and the number of random individuals to add to these.

When the size `n` of the sample is small enough, using `adjust.par=c(0,0,n)` will force the function to
use all individuals. We demonstrate this possibility below for our sample of size 200.

```{r fig.width=10, fig.height=5}
K.low <- lowKi(vcf.file, adjust.par=c(0,0,200))
D.low <- lowKi(vcf.file, adjust.par=c(0,0,200), fraternity = TRUE)
par(mfrow=c(1,2))
plot(K.real, K.low, xlab = "true values", ylab = "LowKi estimates", main = "kinship matrix")
abline(0,1,col="red")
plot(D.real, D.low, xlab = "true values", ylab = "LowKi estimates", main = "fraternity matrix",
     xlim = c(0,0.5), ylim = c(-0.1,0.5))
abline(0,1,col="red")
```

# A second example with a reduced number of individuals

Other files are included in the package, corresponding to a subsample of 20 individuals
taken from the 200 individuals sample described above. Additionnally, an allele frequency 
table, obtained from 100 genotypes individuals from the same population is included as well.
The path to these files is obtained as follows.

```{r}
vcf.file.20 <- system.file("extdata", "shallow_20.vcf.gz", package="LowKi")
kinship.file.20 <-system.file("extdata", "kinship_20.rds", package="LowKi")
fraternity.file.20 <- system.file("extdata", "fraternity_20.rds", package="LowKi")
af.file <- system.file("extdata", "alleleFreqs.tsv", package="LowKi")
```

### Without external allele frequencies

```{r fig.width=10, fig.height=5}
K.low.20 <- lowKi(vcf.file.20)
D.low.20 <- lowKi(vcf.file.20, fraternity = TRUE)

K.real.20 <- readRDS(kinship.file.20)
D.real.20 <- readRDS(fraternity.file.20)

par(mfrow=c(1,2))
plot(K.real.20, K.low.20, xlab = "true values", ylab = "LowKi estimates", main = "kinship matrix"); 
abline(0,1,col="red")
plot(D.real.20, D.low.20, xlab = "true values", ylab = "LowKi estimates", main = "fraternity matrix",
     xlim = c(0,0.5), ylim = c(-0.1,0.5))
abline(0,1,col="red")
```

### Using external allele frequencies

```{r fig.width=10, fig.height=5}
AF <- read.table(af.file, header = TRUE)
K.low.20.af <- lowKi(vcf.file.20, freqs = AF)
D.low.20.af <- lowKi(vcf.file.20, freqs = AF, fraternity = TRUE)

par(mfrow=c(1,2))
plot(K.real.20, K.low.20.af, xlab = "true values", ylab = "LowKi estimates", main = "kinship matrix"); 
abline(0,1,col="red")
plot(D.real.20, D.low.20.af, xlab = "true values", ylab = "LowKi estimates", main = "fraternity matrix",
     xlim = c(0,0.5), ylim = c(-0.1,0.5))
abline(0,1,col="red")
```

