---
title: "LowKi: kinship matrix from low depth data"
author: "Anthony Herzig and Hervé Perdry"
date: "`r Sys.Date()`"
output: html_document
vignette: >
  %\VignetteIndexEntry{LowKi}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r message=FALSE, warning=FALSE, echo = FALSE}
library("knitr")
library("LowKi")
```

LowKi is a package for computing Kinship and Fraternity matrices from shallow sequencing data. It uses vcf files with 'PL' or 'GP' fields. To install the package, use `devtools` :

`devtools::install_github("genostats/LowKi")`

# An example of usage

A small simulated dataset, corresponding to a fictitious small isolated population, is installed with `LowKi`. To build it, five generations of a panmictic population of one thousand individuals were generated by gene dropping; the founders haplotypes were build as mosaics of 1000 genome haplotypes. 

The dataset consists in a 68 Mb vcf file containing shallow data for 200 individuals from the last two simulated generations, with data in 200,000 autosomal SNPs only – to keep the file reasonably small. Two small rds files are also included, corresponding to the 'true' kinship and fraternity matrices, as computed from the length of genome shared IBD by the simulated individuals.

The path to these three files on your filesystem can be obtained as follows:

```{r}
vcf.file <- system.file("extdata", "shallow.vcf.gz", package="LowKi")
kinship.file <-system.file("extdata", "kinship.rds", package="LowKi")
fratern.file <-system.file("extdata", "fraternity.rds", package="LowKi")
```

### Computing the kinship matrix

The function `lowKi` will use the 'PL' field of the vcf file to compute the kinship matrix.

```{r}
K.low <- lowKi(vcf.file)
```

One can also use the hard called genotypes present in the cvf to estimate the kinship matrix (the functions bellow are from package gaston).

```{r}
z <- read.vcf(vcf.file)
K.gen <- GRM(z)
```

Finally, the 'true' kinship values, computed when the dataset was generated, can be loaded by:

```{r}
K.real <- readRDS(kinship.file)
```

Let us compare these values.
```{r fig.width=10, fig.height=5}
par(mfrow=c(1,2))
plot(K.real, K.low, pch = ".", cex = 2, xlab = "true values", 
     ylab = "LowKi estimates")
abline(0,1,col="red")
plot(K.real, K.gen, pch = ".", cex = 2, xlab = "true values",
     ylab = "Estimates from hard called genotypes")
abline(0,1,col="red")
```

### Computing the fraternity matrix

To compute the fraternity matrix (or 'dominance matrix'), just add `fraternity = TRUE` in `lowKi`.

```{r fig.width=10, fig.height=5}
D.low <- lowKi(vcf.file, fraternity = TRUE)
D.gen <- DM(z)
D.real <- readRDS(fratern.file)

par(mfrow=c(1,2))
plot(D.real, D.low, pch = ".", cex = 2, xlab = "true values", 
     ylab = "LowKi estimates", xlim = c(0,0.5), ylim = c(-0.1,0.5))
abline(0,1,col="red")
plot(D.real, D.gen, pch = ".", cex = 2,  xlab = "true values",
     ylab = "Estimates from hard called genotypes", 
     xlim = c(0,0.5), ylim = c(-0.1,0.5)); 
abline(0,1,col="red")
```
